packages/talos_agent/src/agent/errors.rs

```diff
@@ -0,0 +1,55 @@
+use crate::agent::errors::AgentErrorKind::{CertificationTimout, Messaging};
+use crate::agent::model::CertifyRequestChannelMessage;
+use crate::messaging::errors::MessagingError;
+use std::error::Error;
+use strum::Display;
+use thiserror::Error as ThisError;
+use tokio::sync::mpsc::error::SendError;
+
+#[derive(Debug, Display)]
+pub enum AgentErrorKind {
+    Certification { xid: String },
+    CertificationTimout { xid: String, elapsed_ms: u128 },
+    Messaging,
+    Internal,
+}
+
+#[derive(Debug, ThisError)]
+#[error("Talos agent error: '{kind}'.\nReason: {reason}\nCause: {cause:?}")]
+pub struct AgentError {
+    pub kind: AgentErrorKind,
+    pub reason: String,
+    pub cause: Option<Box<dyn Error + Send>>,
+}
+
+impl AgentError {
+    pub fn new_certify_timout(xid: String, elapsed_ms: u128) -> AgentError {
+        AgentError {
+            kind: CertificationTimout { xid, elapsed_ms },
+            reason: "Timeout".to_string(),
+            cause: None,
+        }
+    }
+}
+
+impl From<MessagingError> for AgentError {
+    fn from(e: MessagingError) -> Self {
+        AgentError {
```

Comment:
The `MessageErrorKind` granularity is lost if you want to behave differently for different error kinds.

Reply:
Let's discuss it tomorrow. Not sure how much of low level details client will want to know. Kafka has lots of error codes, I guess we need to pick few critical ones.